File: test/run_isolation_test.py
"""Actual filesystem/decoder checks; fixtures are synthetic, never a model benchmark."""
import hashlib
import json
import os
from pathlib import Path
import shutil
import sys
import tempfile
import time
import unittest
import uuid
sys.path.insert(0, str(Path(__file__).resolve().parents[1] / 'src'))
import media_worker as worker
import media_worker_test as media_fixture


class RunIsolationTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        media_fixture.MediaTests.setUpClass()

    @classmethod
    def tearDownClass(cls):
        media_fixture.MediaTests.tearDownClass()

    def setUp(self):
        self.tmp = tempfile.TemporaryDirectory(prefix='playbook-isolation-test-')
        self.workspace = Path(self.tmp.name)
        self.cwd = Path.cwd()
        os.chdir(self.workspace)
        key = str(uuid.uuid4())
        self.own = {'protocol': 1, 'runId': 'synthetic:' + key, 'sessionId': 's', 'key': key,
                    'workspace': str(self.workspace), 'root': str(self.workspace / '.dsh-runs' / key)}
        receipt = worker.prepare_workspace(self.own)
        self.own.update(prepared=True, realRoot=receipt['root'], markerSha256=receipt['markerSha256'], createdAtMs=receipt['createdAtMs'])
        self.root = Path(receipt['root'])

    def tearDown(self):
        os.chdir(self.cwd)
        self.tmp.cleanup()

    def narration(self, source='brief/narration.txt'):
        (self.root / 'brief/narration.txt').write_text('完整口播。', encoding='utf-8')
        manifest = {'schemaVersion': 1, 'script': source, 'segments': [{'id': 's', 'text': '完整口播。'}]}
        (self.root / 'production.json').write_text(json.dumps(manifest), encoding='utf-8')
        return worker.validate({'kind': 'narration', 'manifest': str(self.root / 'production.json'), 'isolation': self.own}, self.root)

    def test_fresh_allocation_is_idempotent_only_for_same_identity(self):
        result = worker.prepare_workspace(self.own)
        self.assertEqual(result['markerSha256'], self.own['markerSha256'])
        self.assertTrue(all((self.root / name).is_dir() for name in ['brief', 'work', 'evidence', 'artifacts', 'final']))
        other = dict(self.own, runId='another-run')
        with self.assertRaises(worker.Invalid):
            worker.prepare_workspace(other)

    def test_nonempty_arbitrary_root_is_never_adopted(self):
        key = str(uuid.uuid4())
        root = self.workspace / '.dsh-runs' / key
        root.mkdir()
        (root / 'final.mp4').write_bytes(b'old candidate')
        with self.assertRaises(worker.Invalid):
            worker.prepare_workspace(dict(self.own, key=key, root=str(root)))

    def test_current_full_narration_reports_real_host_owner(self):
        result = self.narration()
        self.assertTrue(result['passed'], result)
        self.assertEqual(result['ownership']['runId'], self.own['runId'])
        self.assertTrue(all(b['runId'] == self.own['runId'] for b in result['bindings'].values()))

    def test_old_workspace_manifest_passes_legacy_check_but_not_isolated_check(self):
        (self.workspace / 'old.txt').write_text('完整口播。', encoding='utf-8')
        (self.workspace / 'old.json').write_text(json.dumps({'schemaVersion': 1, 'script': 'old.txt', 'segments': [{'id': 's', 'text': '完整口播。'}]}))
        request = {'kind': 'narration', 'manifest': str(self.workspace / 'old.json')}
        self.assertTrue(worker.validate(request, self.workspace)['passed'])
        result = worker.validate({**request, 'isolation': self.own}, self.root)
        self.assertFalse(result['passed'])
        self.assertIn('CROSS_RUN_PATH', ' '.join(result['failures']))

    def test_foreign_script_under_same_session_workspace_is_rejected(self):
        (self.workspace / 'old-script.txt').write_text('完整口播。')
        result = self.narration(str(self.workspace / 'old-script.txt'))
        self.assertFalse(result['passed'])
        self.assertIn('CROSS_RUN_PATH', ' '.join(result['failures']))

    def test_relative_parent_escape_is_rejected(self):
        result = self.narration('../old-script.txt')
        self.assertFalse(result['passed'])
        self.assertIn('RUN_PATH_TRAVERSAL', ' '.join(result['failures']))

    def test_symlink_and_hardlink_are_not_fresh_outputs(self):
        old = self.workspace / 'old.txt'
        old.write_text('完整口播。')
        (self.root / 'linked.txt').symlink_to(old)
        result = self.narration('linked.txt')
        self.assertIn('RUN_SYMLINK', ' '.join(result['failures']))
        (self.root / 'linked.txt').unlink()
        os.link(old, self.root / 'linked.txt')
        result = self.narration('linked.txt')
        self.assertIn('RUN_HARDLINK', ' '.join(result['failures']))

    def test_changed_model_sidecar_does_not_reassign_the_run(self):
        marker = self.root / '.dsh-run.json'
        data = json.loads(marker.read_text())
        data['runId'] = 'renamed-by-model'
        marker.write_text(json.dumps(data))
        result = self.narration()
        self.assertFalse(result['passed'])
        self.assertIn('RUN_MARKER_CHANGED', ' '.join(result['failures']))

    def test_preserved_old_mtime_is_a_rejection_not_a_creation_attestation(self):
        self.narration()
        source = self.root / 'brief/narration.txt'
        os.utime(source, (time.time() - 86400, time.time() - 86400))
        result = worker.validate({'kind': 'narration', 'manifest': 'production.json', 'isolation': self.own}, self.root)
        self.assertFalse(result['passed'])
        self.assertIn('ARTIFACT_PREDATES_RUN', ' '.join(result['failures']))

    def test_full_media_decode_within_run_and_foreign_video_rejection(self):
        source = media_fixture.MediaTests.root
        files = ['narration.txt', 'title.md', 'captions.srt', 's1.wav', 's2.wav', 'final.mp4', 'pilot.mp4', 'cover.png']
        for name in files:
            shutil.copyfile(source / name, self.root / name)
        manifest = dict(media_fixture.MediaTests.manifest)
        (self.root / 'production.json').write_text(json.dumps(manifest), encoding='utf-8')
        request = {'kind': 'video', 'manifest': 'production.json', 'isolation': self.own}
        result = worker.validate(request, self.root)
        self.assertTrue(result['passed'], result)
        self.assertFalse(result['semanticVerification'])
        # Technical validation cannot tell an unknown copy from new bytes. Engine hash history handles known duplicates.
        self.assertEqual(result['video']['binding']['sha256'], hashlib.sha256((source / 'final.mp4').read_bytes()).hexdigest())
        manifest['video'] = str(source / 'final.mp4')
        (self.root / 'production.json').write_text(json.dumps(manifest), encoding='utf-8')
        result = worker.validate(request, self.root)
        self.assertFalse(result['passed'])
        self.assertIn('CROSS_RUN_PATH', ' '.join(result['failures']))

File: test/run-isolation.test.js
import test from 'node:test'
import assert from 'node:assert/strict'
import { join } from 'node:path'
import { IsolatedPlaybookEngine, callerFacts } from '../src/run-isolation.js'
import { createIsolationManager } from '../src/host-isolation.js'
import { createMediaRunner } from '../src/host-media.js'
const book = { id: 'video-fixture', stages: [
  { id: 'script', objective: 'Freeze script', gate: { evidence: [{key:'production_manifest',type:'string'}], validators: [{kind:'narration',pathKey:'production_manifest'}] } },
  { id: 'qa', objective: 'Check media', gate: { evidence: [{key:'production_manifest',type:'string'}], validators: [{kind:'video',pathKey:'production_manifest'}] }, retry:{ maxAttempts:2 }, next:null },
] }
const caller = (id, workspace='/workspace/test') => ({ agent:{ id, session:{header:{cwd:workspace}, requestHeader:()=>({config:{provider:'test-provider',model:'test-model'}})} }, signal:new AbortController().signal, token:Symbol(),callId:'tool-call' })
function engine() { const e = new IsolatedPlaybookEngine(); e.register(book); return e }
async function prepare(e,id) {
  const own=e.status(id).isolation
  await e.markPrepared(id,own.runId,{status:'pass',runId:own.runId,key:own.key,root:own.root,markerSha256:'1'.repeat(64),createdAtMs:1234})
  return e.status(id).isolation
}
function result(own,kind,digest='2'.repeat(64)) {
  const file={path:join(own.root,'final','final.mp4'),sha256:digest,bytes:16}
  return {validatorVersion:'0.4.0',kind,status:'pass',passed:true,failures:[],
    ownership:{runId:own.runId,key:own.key,root:own.realRoot,markerSha256:own.markerSha256,createdAtMs:own.createdAtMs},
    bindings:{[file.path]:file},narration:{scriptSha256:'script',segmentSha256:'segments'},
    ...(kind==='video'?{video:{binding:file,durationSeconds:4},cover:{sha256:'cover'}}:{})}
}
async function started(e,id) {e.noteCaller(caller(id));await e.start(id,'video-fixture',{});return prepare(e,id)}
async function script(e,id,own) {return e.submit(id,{stageId:'script',evidence:{production_manifest:'production.json'},runtimeChecks:[result(own,'narration')]})}

test('same project and same output request allocate distinct engine-owned roots',async()=>{
  const e=engine();const a=await started(e,'a'),b=await started(e,'b')
  assert.notEqual(a.root,b.root);assert.equal(a.workspace,b.workspace);assert.notEqual(a.runId,b.runId)
  assert.equal(e.status('a').isolation.modelRouteAtStart.model,'test-model')
})
test('input run_id/model_id/workspace cannot forge allocation or producer metadata',async()=>{
  const e=engine();e.noteCaller(caller('a'));await e.start('a','video-fixture',{run_id:'old',model_id:'pretend-model',workspace:'/old'})
  const own=e.status('a').isolation;assert.notEqual(own.runId,'old');assert.equal(own.workspace,'/workspace/test');assert.equal(own.modelRouteAtStart.model,'test-model')
})
test('missing Host cwd fails closed for media, but ordinary source-code tasks are unaffected',async()=>{
  const e=engine();await assert.rejects(e.start('a','video-fixture',{}),/ISOLATION_CONTEXT_MISSING/)
  e.register({id:'coding',stages:[{id:'inspect',objective:'Inspect existing source'}]});await e.start('a','coding',{});assert.equal(e.status('a').isolation,null)
})
test('run cannot submit before preparing its allocated directory',async()=>{
  const e=engine();e.noteCaller(caller('a'));await e.start('a','video-fixture',{})
  await assert.rejects(e.submit('a',{stageId:'script'}),/WORKSPACE_NOT_PREPARED/)
})
test('runtime ownership, not evidence sidecar run_id, determines acceptance',async()=>{
  const e=engine(),a=await started(e,'a'),b=await started(e,'b')
  const status=await e.submit('a',{stageId:'script',evidence:{run_id:a.runId,production_manifest:'production.json'},runtimeChecks:[result(b,'narration')]})
  assert.equal(status.lastGate.passed,false);assert.match(status.lastGate.failures.join(' '),/RUN_OWNERSHIP_MISMATCH/)
})
test('a forged current owner does not make a foreign artifact path belong to this run',async()=>{
  const e=engine(),a=await started(e,'a'),r=result(a,'narration');r.bindings={old:{path:'/workspace/test/old/final.mp4',sha256:'2'.repeat(64),bytes:16}}
  const out=await e.submit('a',{stageId:'script',evidence:{production_manifest:'production.json'},runtimeChecks:[r]});assert.match(out.lastGate.failures.join(' '),/CROSS_RUN_PATH/)
})
test('byte-identical final from another run is rejected even at a new path',async()=>{
  const e=engine(),a=await started(e,'a');await script(e,'a',a);assert.equal((await e.submit('a',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(a,'video')]})).lastGate.passed,true)
  const b=await started(e,'b');await script(e,'b',b)
  const out=await e.submit('b',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(b,'video')]})
  assert.equal(out.lastGate.passed,false);assert.match(out.lastGate.failures.join(' '),/CROSS_RUN_DUPLICATE/)
})
test('legacy retained candidate hashes are checked without relabelling the old run',async()=>{
  const e=engine();e.hydrate({runs:{old:{id:'legacy-id',sessionId:'old',state:'completed',candidates:[{artifacts:{video:{binding:{path:'/old/final.mp4',sha256:'2'.repeat(64)}}}}]}}})
  const a=await started(e,'a');await script(e,'a',a);const out=await e.submit('a',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(a,'video')]})
  assert.match(out.lastGate.failures.join(' '),/CROSS_RUN_DUPLICATE/);assert.equal(e.runs.get('old').isolation,undefined)
})
test('same-run controlled repair preserves root, model facts and budget history',async()=>{
  const e=engine(),a=await started(e,'a');await script(e,'a',a)
  await e.repair('a','script','A concrete correction of the current full narration')
  assert.equal(e.status('a').isolation.root,a.root);assert.equal(e.status('a').run.selfRepairs,1)
})
test('known outputs from archived runs survive restarts',async()=>{
  const e=engine(),a=await started(e,'a');await script(e,'a',a);await e.submit('a',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(a,'video')]})
  await e.start('a','video-fixture',{});await prepare(e,'a')
  const restored=engine();restored.hydrate(e.snapshot());assert.equal(restored.knownOutputs('new-run').some(f=>f.runId===a.runId),true)
  assert.equal(restored.status('a').isolation.prepared,true)
})
test('user-registered prior digest blocks untracked legacy output',async()=>{
  const e=engine();await e.excludeHash(caller('a'),'2'.repeat(64));const a=await started(e,'a');await script(e,'a',a)
  const out=await e.submit('a',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(a,'video')]});assert.match(out.lastGate.failures.join(' '),/EXCLUDED_OUTPUT_HASH/)
})
test('new unrelated output passes and lineage is explicitly validation-time not generation attestation',async()=>{
  const e=engine(),a=await started(e,'a');await script(e,'a',a);await e.submit('a',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(a,'video')]})
  const report=e.report('a');assert.ok(report.artifactLineage.length);assert.equal(report.artifactLineage[0].creationAttested,false)
  assert.equal(report.artifactLineage[0].runId,a.runId);assert.match(report.provenanceLimit,/NOT proof/)
})
test('preparation and path guards use original Host pipeline, not a mutable global cwd',async()=>{
  const e=engine(),exec=caller('a');e.noteCaller(exec);await e.start('a','video-fixture',{})
  let manager,dispatched
  const ctx={tools:{execute:async input=>{
    dispatched=input;assert.equal(manager.isPreparation(input),true);assert.equal(manager.guard(input),undefined)
    const own=e.status('a').isolation
    return {isError:false,value:{kind:'foreground',exitCode:0,signal:null,aborted:false,timedOut:false,stdout:{truncated:false,text:JSON.stringify({status:'pass',runId:own.runId,key:own.key,root:own.root,createdAtMs:1234,markerSha256:'1'.repeat(64)})}}}
  }}}
  manager=createIsolationManager(ctx,e);const out=await manager.prepare(exec)
  assert.equal(dispatched.parent,exec.token);assert.equal(dispatched.agent,exec.agent);assert.equal(dispatched.arguments.workdir,'/workspace/test')
  assert.equal(exec.agent.session.header.cwd,'/workspace/test')
  assert.match(manager.guard({...exec,name:'write',arguments:{file_path:'/workspace/test/final.mp4'}}),/RUN_OUTPUT_PATH/)
  assert.equal(manager.guard({...exec,name:'write',arguments:{file_path:join(out.workspace,'final','title.md')}}),undefined)
  assert.match(manager.guard({...exec,name:'bash',arguments:{command:'echo hi'}}),/RUN_WORKDIR_REQUIRED/)
  assert.equal(manager.guard({...exec,name:'bash',arguments:{command:'echo hi',workdir:out.workspace}}),undefined)
  assert.match(manager.guard({...exec,name:'read',arguments:{file_path:'/workspace/test/.dsh-runs/other/final.mp4'}}),/CROSS_RUN_READ/)
  assert.equal(manager.guard({...exec,name:'read',arguments:{file_path:'/srv/public-tools/README.md'}}),undefined)
  assert.match(manager.guard({...exec,name:'write',arguments:{file_path:join(out.workspace,'.dsh-run.json')}}),/RUN_MARKER_PROTECTED/)
})
test('denied preparation leaves the run unprepared, no fallback side effect',async()=>{
  const e=engine(),exec=caller('a');e.noteCaller(exec);await e.start('a','video-fixture',{})
  const manager=createIsolationManager({tools:{execute:async()=>({isError:true})}},e)
  await assert.rejects(manager.prepare(exec),/denied/);assert.equal(e.status('a').isolation.prepared,false)
})
test('Host validator command receives frozen run ownership, never evidence ownership',async()=>{
  const e=engine(),own=await started(e,'a'),exec=caller('a');let called
  const runner=createMediaRunner({tools:{execute:async value=>{called=value;return {isError:true}}}},e)
  await runner([{kind:'narration',pathKey:'production_manifest'}],{production_manifest:'production.json',isolation:{runId:'fake'}},exec)
  assert.equal(called.arguments.workdir,own.root);assert.equal(called.parent,exec.token)
  assert.ok(!called.arguments.command.includes('fake'))
})
test('failed state persistence does not leave an allocated run in memory',async()=>{
  const e=new IsolatedPlaybookEngine({persist:async()=>{throw new Error('disk full')}});e.register(book);e.noteCaller(caller('a'))
  await assert.rejects(e.start('a','video-fixture',{}),/disk full/);assert.equal(e.status('a').run,undefined)
})

test('concurrent identical candidates cannot both commit as independent runs',async()=>{
  const e=engine(),a=await started(e,'a'),b=await started(e,'b');await script(e,'a',a);await script(e,'b',b)
  const outcomes=await Promise.allSettled([
    e.submit('a',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(a,'video')]}),
    e.submit('b',{stageId:'qa',evidence:{production_manifest:'production.json'},runtimeChecks:[result(b,'video')]})
  ])
  const committed=[e.status('a'),e.status('b')].filter(row=>row.run.state==='completed')
  assert.equal(committed.length,1)
  assert.ok(outcomes.some(row=>row.status==='rejected'||row.value?.lastGate?.passed===false))
})
